driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -652,129 +561,135 @@ public ByteBuf getBuffer(final int size) {
-                    try {
-                        MessageHeader messageHeader = new MessageHeader(headerInputBuffer, description.getMaxMessageSize());
-                        replyHeader = new ReplyHeader(headerInputBuffer, messageHeader);
-                    } finally {
-                        headerInputBuffer.close();
-                    }
-
-                    if (replyHeader.getMessageLength() == TOTAL_REPLY_HEADER_LENGTH) {
-                        onSuccess(new ResponseBuffers(replyHeader, null));
-                    } else {
-                        readAsync(replyHeader.getMessageLength() - TOTAL_REPLY_HEADER_LENGTH,
-                                  new ResponseBodyCallback(replyHeader));
-                    }
-                } catch (Throwable t) {
-                    close();
-                    callback.onResult(null, t);
-                }
-            }
-        }
-
-        private void onSuccess(final ResponseBuffers responseBuffers) {
-
-            if (responseBuffers == null) {
-                callback.onResult(null, new MongoException("Unexpected empty response buffers"));
-                return;
-            }
-
+        public void onResult(final ByteBuf result, final Throwable t) {
             try {
-                callback.onResult(responseBuffers, null);
-            } catch (Throwable t) {
-                LOGGER.warn("Exception calling callback", t);
+                if (t != null) {
+                    throw t;
+                }
+                MessageHeader messageHeader = new MessageHeader(result, description.getMaxMessageSize());
+                readAsync(messageHeader.getMessageLength() - MESSAGE_HEADER_LENGTH, new MessageCallback(messageHeader));
+            } catch (Throwable localThrowable) {
+                callback.onResult(null, localThrowable);
+            } finally {
+                if (result != null) {
+                    result.release();
+                }
             }
         }
 
-        private class ResponseBodyCallback implements SingleResultCallback<ByteBuf> {
-            private final ReplyHeader replyHeader;
+        private class MessageCallback implements SingleResultCallback<ByteBuf> {
+            private final MessageHeader messageHeader;
 
-            ResponseBodyCallback(final ReplyHeader replyHeader) {
-                this.replyHeader = replyHeader;
+            MessageCallback(final MessageHeader messageHeader) {
+                this.messageHeader = messageHeader;
             }
 
             @Override
             public void onResult(final ByteBuf result, final Throwable t) {
-                if (t != null) {
-                    try {
-                        callback.onResult(new ResponseBuffers(replyHeader, result), t);
-                    } catch (Throwable tr) {
-                        LOGGER.warn("Exception calling callback", tr);
+                try {
+                    if (t != null) {
+                        throw t;
```

Comment:
Using Exceptions as control flow again seems wrong / counter to how we have handled throwables elsewhere with callbacks.